OICI 180:Fix-code unicity

Iandraina-etech opened the pull request:
Fixed code unicity for:
Speciality 
Prescriber 

Now the code isn't unique on the database instead the FE send a request to the BE to check if the code already exist on a valid entity

Also changed the right to admin user rights on openimis (equals  to the right user from openimis core module)

Iandraina-etech committed:
Change the way code unicity is implemented

changed the way code unicity is implemented in the backend section of the code for the class Speciality and Prescriber

Iandraina-etech committed:
Matched right

changed Right to Match admin openimis rights on the backend

Iandraina-etech committed:
Merge pull request #26 from ArkeupIDComores/check-filter-implementation

Fix-unicity code

Iandraina-etech committed:
Merge branch 'mngoe:dev-comores' into dev-comores

kevel-dev commented on claim/apps.py:
```diff
@@ -4,6 +4,7 @@
 
 DEFAULT_CFG = {
     "default_validations_disabled": False,
+    "gql_query_admin_openimis_perms": ["121701"],
```
A priori ces changes dans les permissions n'ont rien a voir avec l'unicité du code et ne doivent pas etre présent

Iandraina-etech replied:
On a du changer toutes ces permissions car normalement ce sont les admin openimis qui ont ce droit et non pas les contributeurs de claim , ce droit est l' equivalent du right user du module core que seuls les admin openimis ont

kevel-dev replied:
Vous avez changé délibérément ou bien c'est marqué dans les RFC ? Si ce n'est pas dans les RFC il faut simplement remonter le probleme chez Mr Alim Zé pour évaluation et faire un autre ticket séparé pour correction.

Iandraina-etech replied:
dans le OICI il dit que c'est l administrateur openimis qui a les droits d'insertion et d'update et delete 
Au debut on pensait que l'admin openimis avait aussi les droits de l'administrateur claims sauf qu'il s'avere que ce n'est pas le cas (cela a ensuite causé un probleme sur le droit de delete ,voila pourquoi j'ai modife cette partie du code ,c'est donc un changement requis par l'oici)
<img width="801" height="161" alt="image_2025-10-27_144630880" src="https://github.com/user-attachments/assets/a5075a8b-6d7f-43d6-9efa-2fbc90c1845c" />

kevel-dev replied:
Cela veut dire qu'un simple administrateur de claim ne pourra plus faire ces traitements. En gros seule l'admin pourra le faire, normalement les autres users devraient continuer a faire ces actions comme par le passé en plus de l'admin openIMIS.

Iandraina-etech replied:
Oui et c'est deja le comportement de l'appli actuellement dans le front-end le menu d'acces de prescriber et speciality est protegé par right user ,seuls les admin openimis peuvent effectuer ces actions c'est conforme au specification du ticket seuls les admin openimis peuvent faire ces action 
comme precisé dans le ticket 
 Je pense que ces corrections sont necessaire pour etre conforme a l'oici 

<img width="1840" height="881" alt="image_2025-10-27_155409669" src="https://github.com/user-attachments/assets/498e215a-dd3d-4eaa-b683-00d81bb52e79" />

kevel-dev replied:
Sauf que la vous avez fait un change en back sur l'objet claim en entier et non sur le champ prescripteur de cet objet.

Iandraina-etech replied:
Justement je n'ai changé que les permissions des objets Prescripteur et Spécialités et non pas les champs dans claim(ceux la sont seulement accessible que par l'admin openimis), ces changements n'affectent pas l'objet Claim mais seulement les nouvelles entités donc je ne vois pas où est le problème

kevel-dev commented on claim/gql_mutations.py:
```diff
@@ -1073,7 +1073,7 @@ class Input(CreatePrescriberInputType):
     def async_mutate(cls, user, **data):
         try:
-            if not user.has_perms(ClaimConfig.gql_mutation_create_claims_perms):
+            if not user.has_perms(ClaimConfig.gql_query_admin_openimis_perms):
```
Meme commentaire que celui du haut, tout ces changes sur les permissions ne doivent pas etre présent.

kevel-dev commented on claim/services.py:
```diff
@@ -828,6 +836,13 @@ def set_deleted(self, prescriber):
+    
+    @staticmethod
+    def check_unique_speciality_code(code):
+        if Speciality.objects.filter(code=code, validity_to__isnull=True).exists():
```
Il faut retirer la validité de telle sorte que meme si la spécialité c'est plus valide le code doit toujours etre unique. Sinon ca risque embrouiller lors des debugs avec les recherches SQL. on aura plusieurs enrégistrements avec le meme code, ca doit etre unique peut importe la validité.

kevel-dev commented on claim/services.py:
```diff
@@ -758,6 +758,14 @@ class PrescriberService:
 
+    @staticmethod
+    def check_unique_prescriber_code(code):
+        if Prescriber.objects.filter(code=code, validity_to__isnull=True).exists():
```
Il faut retirer la validité de telle sorte que meme si la spécialité c'est plus valide le code doit toujours etre unique. Sinon ca risque embrouiller lors des debugs avec les recherches SQL. on aura plusieurs enrégistrements avec le meme code, ca doit etre unique peut importe la validité.

Iandraina-etech committed:
Update services.py

Removed validity to  is null poure que le code soit unique meme si on efface certains objets le code reste unique
